repo.js

```diff
@@ -1,81 +1,200 @@
-  this.privateKey = this.archive.privateKey
-  var signalhub = Signalhub('dat-' + this.archive.key.toString('hex'), opts.signalhub || 'https://signalhub.mafintosh.com')
-  this.swarm = this.swarm || swarm(signalhub)
-  self.join()
-  this._open(key)
-}
+const EventEmitter = require('events').EventEmitter
+const Signalhub = require('signalhubws')
+const hyperdrive = require('hyperdrive')
+const ram = require('random-access-memory')
+const websocket = require('websocket-stream')
+const WebrtcSwarm = require('webrtc-swarm')
+const pump = require('pump')
+var through = require('through2')
 
-inherits(Repo, events.EventEmitter)
+const DEFAULT_WEBSOCKET_RECONNECT = 1000
+const DEFAULT_WEBSOCKET_CONNECTION_DELAY = 1000
+const DEFAULT_WEBSOCKET_CONNECTION_DELAY_LONG = 5000
+const DAT_PROTOCOL = 'dat://'
 
-Repo.prototype._open = function () {
-  var self = this
-  this.archive.open(function () {
-    self.emit('ready')
-  })
-}
+const DEFAULT_SIGNALHUBS = ['ws://gateway.mauve.moe:3300']
 
-/**
- * Joins the swarm for the given repo.
- * @param  {Repo}   repo
- */
-Repo.prototype.join =
-Repo.prototype.resume = function () {
-  this.swarm.on('peer', this._replicate.bind(this))
-}
+// Check if the page was loaded from HTTPS
+const IS_SECURE = self.location.href.startsWith('https')
 
-/**
- * Internal function for replicating the archive to the swarm
- * @param  {[type]} peer A webrtc-swarm peer
- */
-Repo.prototype._replicate = function (conn) {
-  var peer = this.archive.replicate({
-    upload: true,
-    download: true
-  })
-  pump(conn, peer, conn)
-}
+module.exports =
 
-/**
- * Leaves the swarm for the given repo.
- * @param  {Repo}   repo
- */
-Repo.prototype.leave =
-Repo.prototype.pause = function () {
-  this.swarm.removeListener('peer', this._replicate)
-}
+class Repo extends EventEmitter {
+  /**
+  * A dat repository is a hyperdrive with some default settings.
+  * @param {string} url    The url
+  * @param {Object} opts   Options to use in the archive instance
+  */
+  constructor (url, opts) {
+    if(!opts) throw new TypeError('Repo must have optsions passed in from `Dat` instance')
+    super()
+    let key = null;
 
-Repo.prototype.destroy =
-Repo.prototype.close = function () {
-  var self = this
-  self.swarm.close(function () {
-    self.archive.close(function () {
-      self.db.close(function () {
-        self.emit('close')
+    // Make sure that the URL looks like `dat://somekey` if it exists
+    if(url) {
+      if(url.indexOf(DAT_PROTOCOL) === 0) {
+        key = url.slice(DAT_PROTOCOL.length)
+      } else {
+        key = url
+        url = DAT_PROTOCOL + key
+      }
+    }
+
+    this.url = url
+    this.opts = opts
+    this.db = this.opts.db || ram
+    this.archive = hyperdrive(this.db, key, opts)
+    this._isReady = false
+
+    this.signalhub = null
+    this.swarm = null
+    this.websocket = null
+    this._websocketTimer = null
+
+    // If no URL was provided, we should set it once the archive is ready
```

Comment:
I wonder if this can happen in the _open function which sets the `_isReady` bit that `this.ready` is waiting for

Reply:
Good call!